ISSUE-524: better native solr date range handling, better date math/timezone and also stricter checks on User input

DiegoPino opened the pull request:
# What?

See #524 and follow us on slack for more "dating" tips 😮‍💨 

Still missing here:
- ~add BCE/AC select box to the year/full date entry boxes in the widget~. Done differently. Full date input hides if one of the ranges is negative. Less UI for now is good.
- ~Totally remake the JS side of things date parsing bc the timestamps for BCE are totally wrong (no clue yet how)~ DONE
- ~Check/double check all Widget settings we have. See if all are respected.~ DONE (the old range widget was not working. mea culpa. Fixed)
- ~Incorporate a few shared ideas in slack like  allow "extra slack/+-gap" to any start/end when faceting via JSON on the backend, that way even if the query is 2021-2025(my most complex years)~. So far the solution is approximate but good enough. The last gap now uses the modulus of the GAP (bc it will not be exact) and we pass that data between query and widget back and forth. I can get an histogram that returns 2018-2028 (with 2028  maybe a good year?)

In the case of a single range (1899-01-01 to December 31st of the same year) the histogram turns into a single bar, If not it becomes invisible. 

Will follow up with more pulls for that tomorrow

DiegoPino committed:
Ensure that the Time ZONE offset is always a full INT. Why?

Because Time Zones play strange with BCE dates. And the initial offset of time for a date like -50 BCE is 04 and fraction hours (and then the comes the time zone added)

DiegoPino committed:
Because of Solr Native Date Range v/s Timestamp based normal date, we need to normalize faceted results

Adding  public static function DateToUnix($date). Called bu All he Widgets. I can't use DATE_ATOM here though (don't ask me why.. PHP refuses to consume its own format for BCE, but a plain call to \DateTimeImmutable results in proper BCE timestamp

DiegoPino committed:
Normalize ::getRangeFromResults() so it can handle both native Solr date range results and timestamp based

DiegoPino committed:
Date math. Re-apply the DateTime Offset generated by the timezone. Index timezone, user timezone, time-zone less unixtimestamp

Better year logic/start/end and more checks for data validation

patdunlavey commented:
@DiegoPino I'm going to be out of town until Monday. I'll review then if you still need.

DiegoPino commented:
@patdunlavey enjoy the weekend. Let's do a rain check on Monday. Thanks

DiegoPino committed:
just normalize the Title so it does not look weird

DiegoPino commented:
@patdunlavey. I believe I have to keep further refining this pull. It does work (at least better than before) and native Solr date ranges work too, but I have the feeling the cut-off/end range is being reduced more than I would like and there might be a floating point (or any other math) issues lingering still or it might be the cut-off settings of the ranged faceting sent to Solr too. Debugging is required to be sure. Also, if you test please fetch a fresh strawberryfield 1.5.0 bc you might need it if you decide to go for BCE dates.

I can come back to this on Friday (have to travel tomorrow). Thanks a lot

DiegoPino committed:
Improve Slider JS to deal with BCE in multiple browsers

Figure out that Chrome does not parse BCE, but Safari does. This is the first pass. Will add the Select BCE/AC too in a second commit.

I do wonder now if allowing the last second/last minute of a year is what people expect or not

DiegoPino committed:
Why JS? Why make months from 0 to 11? You are not even a real programing language!

DiegoPino committed:
Ok! Much better. But. I still need to deal with a single year offset at the end, depending on the cut-off settings

Leaving for today my error_logs() bc i need to keep debugging tomorrow.

DiegoPino committed:
Note about removing Time Zone offset from Facet results.

What? We add the offset (or remove) during query to match "input" v/s "actual indexed data". All good there, but once we get the FACETs back, those ranged/buckets have the offsets. So if we use those as input again and YET again calculate offsets, we are basically moving hours back and forth and WE CAN (on the $max) end escaping/reducing our actual range.
Solution for the weekend is re-verse the operation we do on query, so our facets are YET again UTC

DiegoPino committed:
Unsure what happened here. But this was making the old range (that worked) fail. Should be fixed now

DiegoPino committed:
Reusing the settings from the parent class for the processor breaks Jquery validation

DiegoPino committed:
Use modulus to approximate the width of the last range. Pass it as a Widget override

@patdunlavey @alliomeria when using dynamic gaps and ceiling it, the last range might be shorter than the calculate range. In practice, this does not affect the proper "filtering" which is always exact, but will show, in very long start/end dates an end range that expands way over the actual data covered. The way Solr generates the gaps by default is un-even to ensure the start/end dates are covered, but does not mean that we will have data for the complete range. This is of course a performance need, we can't request 1000 facets with static gaps of 1 year. But with this at least we can ensure than the "histogram", even if covering more "ground" that the actual data is cut at actual gap size used in the backend and not, e.g the 20 YEARS used for every other gap in the response.

I will remove the error_logs once i have tested more. This is way more robust now, even if the histogram might still show data that escapes by an smaller error (which i can't avoid, it is how Solr works) on the upper side.

DiegoPino committed:
Hide Full Date entry if $min year is negative.

Work around for now. Might evaluate in 1.6.0 if it really makes sense adding more UI for BCE dates at all. Repositories are already too complicated for end users

DiegoPino committed:
Explain settings better inline. Also make sure $max with timezone offset is used.

DiegoPino committed:
Clean up JS/remove console.log messages

DiegoPino committed:
Make chart a bar if data is a single value. If not it becomes invisible

DiegoPino commented:
@patdunlavey I will wait until the end of the day for comments on this. If not I will merge because as it is right now it is a real improvement over the previous logic. Thanks!

DiegoPino committed:
Remove error logs and preparing to a later night merge!
